File: RayTracerAvalonia/RayTracing/Shapes/Sphere.cs
using RayTracerAvalonia.RayTracing.Extensions;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Numerics;
using System.Runtime.CompilerServices;
using Avalonia.Controls.Documents;

namespace RayTracerAvalonia.RayTracing.Shapes;
public readonly struct Sphere(Vector3 center, float radius, IMaterial appearance) : IShape
{





    public Vector3 Center { get; } = center;
    public float Radius { get; } = radius;

    public IMaterial Appearance => appearance;

    public Vector3 GetNormalAt(Vector3 point)
    {
        // Calculate the normal vector at a point on the sphere
        var normal = point - Center;
        return normal.Normalize();
    }

    private readonly List<float> _emptyList = [];

    [MethodImpl(MethodImplOptions.AggressiveInlining)]
    public List<float> Intersect(Ray ray)
    {
        var os = ray.Start - Center;
        var b = 2 * Vector3.Dot(os, ray.Direction);
        var c = os.LengthSquared() - Radius * Radius;
        var disc = b * b - 4 * c;
        
        switch (disc)
        {
            case < 0:
                return _emptyList;
            case 0:
                return [-b / 2];
            default:
                {
                    var root = Math.Sqrt(disc);
                    return [(float)((-b - root) / 2f), (float)((-b + root) / 2f)];
                }
        }
        
    }


}

File: RayTracerAvalonia.Desktop/cpp_raytracing_engine/sphere.cpp
#include "include/sphere.hpp"
#include <cmath>
#include <limits>

auto Sphere::getColor() const -> Color { return color; }

auto Sphere::intersect(const Ray &ray) const noexcept -> std::vector<float> {
  Vector3f os = ray.start - center;
  float b = 2.0f * os.dot(ray.direction);
  float c = os.squaredNorm() - radius * radius;
  float disc = b * b - 4.0f * c;

  if (disc < 0) {
    // No intersection
    return {};
  } else if (disc == 0) {
    // Tangent intersection
    return {-b / 2.0f};
  } else {
    // Two intersections
    float root = std::sqrt(disc);
    float t1 = (-b - root) / 2.0f;
    float t2 = (-b + root) / 2.0f;
    return {t1, t2};
  }
}

auto Sphere::getNormalAt(Vector3f point) const noexcept -> Vector3f {
  // Calculate the normal vector at a point on the sphere
  Vector3f normal = point - center;
  return normal.normalized();
}
